history/src/block_stores/multiple_strategy_block_store.rs

```diff
@@ -75,67 +86,113 @@ impl BlockStorageInterface for MultipleStrategyBlockStorage {
-            if persistent_block_range.contains(&slot) {
-                self.persistent_block_storage.get(slot, config).await
-            } else if let Some(faithful_rpc_client) = self.faithful_rpc_client.clone() {
-                match faithful_rpc_client
-                    .get_block_with_config(slot, config)
-                    .await
-                {
-                    Ok(block) => Ok(ProducedBlock::from_ui_block(
-                        block,
-                        slot,
-                        CommitmentConfig::finalized(),
-                    )),
-                    Err(_) => bail!(BLOCK_NOT_FOUND),
+            false => {
+                debug!(
+                    "Block {} not found in persistent block-storage - continue",
+                    slot
+                );
+            }
+        }
+
+        if let Some(faithful_block_storage) = &self.faithful_block_storage {
+            match faithful_block_storage.get_block(slot).await {
+                Ok(block) => {
+                    debug!(
+                        "Lookup for block {} successful in faithful block-storage",
+                        slot
+                    );
+                    return Ok(block);
+                }
+                Err(_) => {
+                    debug!("Block {} not found in faithful storage - giving up", slot);
+                    bail!(format!("Block {} not found in faithful", slot));
                 }
-            } else {
-                bail!(BLOCK_NOT_FOUND);
             }
+        } else {
+            // no faithful available
+            bail!(format!("Block {} not found - faithful not available", slot));
         }
     }
 
-    async fn get_slot_range(&self) -> Range<Slot> {
-        let in_memory = self.inmemory_for_storage.get_slot_range().await;
-        // if faithful is available we assume that we have all the blocks
-        if self.faithful_rpc_client.is_some() {
-            0..in_memory.end
-        } else {
+    // we need to build the slots from right to left; in-memory defines the right most slots
+    async fn get_slot_range(&self) -> RangeInclusive<Slot> {
+        let in_memory_slots = self.inmemory_for_storage.get_slot_range().await;
+
+        // merge them
+
+        let mut lower: Slot = *in_memory_slots.start();
+
+        {
             let persistent_storage_range = self.persistent_block_storage.get_slot_range().await;
-            persistent_storage_range.start..in_memory.end
+            //         x------
+            // |------x
+            if lower - persistent_storage_range.end() <= 1 {
+                // move the lower bound to the left
+                lower = lower.min(*persistent_storage_range.start());
+            }
         }
+
+        if let Some(faithful_block_storage) = &self.faithful_block_storage {
+            let faithful_storage_range = faithful_block_storage.get_slot_range();
```

Comment:
Should remove this case to reduce complexity.

Reply:
if slot < lower we assume that it is in faithful in multi storage side.

Reply:
yes but that is logic of faithful.get_slot_range() IMO